[inputstream] change addon interface to new C++ style

AlwinEsch opened the pull request:
## Description
This commit change the inputstream addon system to the new C++ style
and remove the old parts.

The parts from here are taken from kodi-agile and becomes mainly improved
and optimized from @peak3d on next changes.

Currently it supports only the same as the old, only a new interface way.

<!--- Provide a general summary of your change in the Title above -->

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h:
```diff
@@ -0,0 +1,585 @@
+      SUPPORTSSEEK = (1 << 3),
+
+      /// supports pause
+      SUPPORTSPAUSE = (1 << 4)
```
Maybe it's just personal style, but I would insert an underline between the verb and the action (e.g. SUPPORTS_PAUSE). Increases readability and this is how we usuallly name enum members, iirc.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h:
```diff
@@ -0,0 +1,585 @@
+  {
+    kodi::addon::CInstanceInputStream* addonInstance;
+
+    bool (__cdecl* open)(AddonInstance_InputStream* instance, INPUTSTREAM&);
```
For other addon types, like ADSP, you are using "const pointer to the addon instance", which is imo correct, and you name the parameter <code>addonInstance</code> - would read <code>AddonInstance_InputStream const* addonInstance</code> here. I suggest to be consistent across addon types re naming and data types.

I also suggest to check older commits and to correct this. IIRC, the "const" came in with the ADSP PR, before we did not spot this and always used non-const pointers.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h:
```diff
@@ -0,0 +1,585 @@
+    int (__cdecl* get_time)(AddonInstance_InputStream* instance);
+
+    // IPosTime
+    bool (__cdecl* pos_time)(AddonInstance_InputStream* instance, int);
```
May I suggest to find a more speaking name for this function? "pos_time" means nothing to me.

FernetMenta replied:
It positions the stream to the given time. how much more meanful could it be? :)
this is a demuxer/inputstream interface. addons just implement it and breaking naming in the addon interface does not make sense. you would need to change the entire interace which has a much broader scope than this pr.

ksooo replied:
> It positions the stream to the given time. how much more meanful could it be? :)

 set_stream_position would be MUCH more meaningful, imo.

 Alwin is actually changing the whole interface, all addons must be adapted anyway. So we could and should take the opportunity to improve even small things like this.

 But if anybody else is fine with pos_time feel free to ignore me, I was just suggesting, not demanding.

FernetMenta replied:
not really what it does. you can set position to a position or a time.

AlwinEsch replied:
I leave it, there is only the "C" structure who handle the interface between the both.

There comes the next time a complete reworked addon documentation and there can be the `CInstanceInputStream::PosTime(...)` detailed descripted.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h:
```diff
@@ -0,0 +1,585 @@
+    int64_t (__cdecl* position_stream)(AddonInstance_InputStream* instance);
+    int64_t (__cdecl* length_stream)(AddonInstance_InputStream* instance);
+    void (__cdecl* pause_stream)(AddonInstance_InputStream* instance, double);
+    bool (__cdecl* is_real_time_stream)(AddonInstance_InputStream* instance);
```
I think you should either supply variable names for all parameters or for none. Currently, you supply a name for AddonInstance_InputStream parameter, but none for all other paramters...

Maybe we should also change this for addon types that came in with older PRs, where we missind this inconsistency during review.

ksooo commented on xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:
```diff
@@ -20,40 +20,143 @@
+  std::string name(addonBase->ID());
+
+  m_fileItemProps = StringUtils::Tokenize(listitemprops, "|");
+  for (auto &key : m_fileItemProps)
```
No need for & after auto. Trust the compiler. ;-)

Here and at other places in this source file.

AlwinEsch replied:
No, in this case is needed :D

Has changed them, but there is a write done and was wrote to somewhere else ;)

Montellese replied:
@ksooo you can't trust the compiler because the compiler can't know what you want to do. There are basically three options:
* `for (auto value : values)` which will create a full copy of the object (most often not what you want)
* `for (auto& value : values)` which will create a modifyable reference to the object (similar to `iterator`)
* `for (const auto& value : values` which will create a const reference to the object (similar to `const_iterator`)

This is the same as with regular C++ iterators. You have to tell the compiler whether to use `iterator` or `const_iterator`.

ksooo replied:
@Montellese thanks for explaining. Makes perfect sense.

peak3d replied:
auto & makes sense, if not the loop is working with a copy. Already fall ito this trap

ksooo commented on xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:
```diff
@@ -20,40 +20,143 @@
+    if (addon.asString() != addonBase->ID())
+      return false;
+    else
+      return true;
```
I would write this as <code>return addon.asString() == addonBase->ID();</code>

ksooo commented on xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:
```diff
@@ -20,40 +20,143 @@
-    m_hasPosTime = m_addon->HasPosTime();
-    m_canPause = m_addon->CanPause();
-    m_canSeek = m_addon->CanSeek();
+    memset(&m_caps, 0, sizeof(m_caps));
```
replace `memset` by `m_struct = {{ 0 }};`?

AlwinEsch replied:
In this case not work, there is only one part defined in structure and brought errors.

ksooo replied:
k

AlwinEsch replied:
What mean "k"?

ksooo replied:
it's like "if you say so, although I'm not fully convinced, but I don't really care to investigate"

hudokkow replied:
Means OK.

ksooo replied:
Not exactly, it's more like okay-ish.

ksooo commented on xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:
```diff
@@ -237,29 +357,129 @@ bool CInputStreamAddon::SeekTime(double time, bool backward, double* startpts)
+void CInputStreamAddon::DisposeStreams()
+{
+  for (auto &stream : m_streams)
+    delete stream.second;
```
does the dtor of stream.second take care to delete `ExtraData`, which is allocated here https://github.com/xbmc/xbmc/pull/12325/files#diff-3590db0d5bee056ad4a07281388f0531R455?

AlwinEsch replied:
Yes is in his desctructor

ksooo replied:
Okay.

ksooo reviewed (changes requested):
some questions and suggstions

AlwinEsch committed:
[inputstream] change addon interface to new C++ style

This commit change the inputstream addon system to the new C++ style
and remove the old parts.

The parts from here are taken from kodi-agile and becomes mainly improved
and optimized from @peak3d on next changes.

Currently it supports only the same as the old, only a new interface way.

AlwinEsch commented:
Is updated.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Can someone help me to explain what on Windows now the second time fails? For me not related to this request but I'm not sure.

peak3d commented:
> fatal: repository 'https://github.com/xbmc/FFmpeg/archive/' not found

peak3d commented:
thanx @AlwinEsch for this PR, we already approoved the changes in agile branch and it seems to be mostly identical. 👍 from my side.

Rechi commented:
strange build machine error, disabled it
jenkins build this please

AlwinEsch commented:
jenkins build this please
